cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -309,6 +348,21 @@ std::shared_ptr<ScalarFunction> MakeArithmeticFunctionNotNull(std::string name,
   return func;
 }
 
```

Comment:
Insertion of implicit casts is accomplished by overriding Function::DispatchBest. For example, to ensure that unsigned types are supported by casting to a compatible unsigned type, use:

```suggestion
struct UnaryArithmeticFunction : ScalarFunction {
  using ScalarFunction::ScalarFunction;
  Result<const Kernel*> DispatchBest(std::vector<ValueDescr>* values) const override {
    RETURN_NOT_OK(CheckArity(*values));
    
    using arrow::compute::detail::DispatchExactImpl;
    
    if (auto kernel = DispatchExactImpl(this, *values)) return kernel;
    
    EnsureDictionaryDecoded(values);
    if (auto type = CommonNumeric({values->at(0), int8()})) {
      ReplaceTypes(type, values);
    }
 
    if (auto kernel = DispatchExactImpl(this, *values)) return kernel;
    return arrow::compute::detail::NoMatchingKernel(this, *values);
  }
};
```

(UnaryScalarFunction will replace ScalarFunction below in `auto func = std::make_shared<ScalarFunction>(name, Arity::Unary(), doc);`)

Reply:
Not sure why we need UnaryScalarFunction and can't use ScalarFunction as is. Why the *CommonNumeric* is using *int8()*?

Reply:
`ScalarFunction` does not provide implicit casts, such as from unsigned to signed integers. `UnaryScalarFunction` is provided to add implicit casts including:

```
uint8 -> int16
uint16 -> int32
uint32 -> int64
uint64 -> int64
dictionary<int32, float> -> float
//...
```

The call to CommonNumeric with int8 ensures that the output type is signed, with no more widening than necessary. Insertion of implicit casts is tested for the other arithmetic functions using [CheckDispatchBest](https://github.com/apache/arrow/blob/b4806a0eefb5b21e9f44cac1c7c9113a4c31ad01/cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc#L645)

Reply:
Got it, nice trick!